### PR TITLE
WIP Reduction support.

### PR DESCRIPTION
**Pull request by jsjodin**
This patch enables reduction support for the reduction tests in aomp/trudnk.
It is still a WIP with some debug printouts etc, but it should be able to run the tests successfully.

### CODE REVIEW

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -2231,7 +2231,7 @@ calculateTripCount(Fortran::lower::AbstractConverter &converter,
   };
 
   // Start with signless i32 by default.
-  auto tripCount = b.createIntegerConstant(loc, b.getI32Type(), 1);
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP.cpp:2234]: Just a small nit: Is this change still needed after PR #250 got merged? I think the MLIR trip count should accept any integer type, as long as when emitting the call to `__tgt_target_kernel` it gets casted to i64 to match the function signature. That was something that got introduced with that PR (in `emitTargetCall()` in line 5201 of OMPIRBuilder.cpp).

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 2]
```diff
@@ -2128,18 +2109,36 @@ convertOmpDistribute(Operation &opInst, llvm::IRBuilderBase &builder,
...
+    ompBuilder->RIManager.setPrivateVarAllocaIP(allocaIP);
+
+    auto regionBlock =
+        convertOmpOpRegions(opInst.getRegion(0), "omp.distribute.region",
+                            builder, moduleTranslation, bodyGenStatus);
+
+    builder.SetInsertPoint(regionBlock->getTerminator());
+
+    // FIXME(JAN): We need to know if we are inside a distribute and
+    // if there is an inner wsloop reduction, in that case we need to
```
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2121]: Not sure if this helps here, but a method called `getInnermostCapturedOmpOp()` was added to omp.target, which you could get a reference to via `opInst` parents. If that omp.target op contains a multi-level nest of OpenMP ops with the innermost region corresponding to an omp.wsloop, it will return that op. I guess you could read its reduction clause at that point, if needed. Something like:
```c++
if (auto targetOp = opInst->getParentOfType<omp::TargetOp>()) {
  if (auto wsLoopOp = dyn_cast_if_present<omp::WsLoopOp>(targetOp.getInnermostCapturedOmpOp())) {
    if (wsLoopOp.getReductions()...) { ... }
  }
}
```
Also, the `TargetOp::isTargetSPMDLoop()` function will tell whether it is a "target teams distribute parallel do", if you need checking this here.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp [hunk 3]
```diff
@@ -2718,7 +3817,11 @@ getKmpcForStaticLoopForType(Type *Ty, OpenMPIRBuilder *OMPBuilder,
 static void createTargetLoopWorkshareCall(
     OpenMPIRBuilder *OMPBuilder, WorksharingLoopType LoopType,
     BasicBlock *InsertBlock, Value *Ident, Value *LoopBodyArg,
-    Type *ParallelTaskPtr, Value *TripCount, Function &LoopBodyFn) {
+    Type *ParallelTaskPtr, Value *TripCountOrig, Function &LoopBodyFn) {
+  // FIXME(JAN): The trip count is 1 larger than it should be, this may not be
+  // the right way to fix it, but it could be.
+  Value *TripCount = OMPBuilder->Builder.CreateSub(
```
Comment (skatrak) [inline llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:3823]: Does this change break regular "target teams distribute parallel do" in exchange for making it work with reductions, or can the off-by-one problem be reproduced in both situations?

**Review by skatrak — COMMENTED** [review]
I just have some small questions and suggestions, trying mostly to help somewhat. I haven't looked too closely into the whole approach, since I guess it's still bound to change a bit.

**Comment by jsjodin** [comment]
Rebased with latest ATD.

File: flang/lib/Lower/OpenMP.cpp (hunk 1 above)
Comment (jsjodin) [reply flang/lib/Lower/OpenMP.cpp]: I was able to remove this change. Thanks for pointing it out.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp (hunk 2 above)
Comment (jsjodin) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2119]: Perhaps, I will look into this later on.

File: llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp (hunk 3 above)
Comment (jsjodin) [reply llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp]: It doesn't seem to affect target teams distribute parallel to, but it is clear that the DeviceRTL functions for reductions want the reduced trip count.  I would be surprised if the other DeviceRTL functions would assume a different trip count.
